### PR TITLE
Initial MFA support for SPO, PnP, Teams, MSOnline and AzureAD

### PR DESCRIPTION
**Pull request by brianlala**
#### Pull Request (PR) description
This release enables Office365DSC to leverage the dependent external MSCloudLoginAssistant module to streamline authentication to Office 365 platforms, and provide support for multi-factor authentication for most workloads. Exchange and Security & Compliance are the two current exceptions.

#### This Pull Request (PR) fixes the following issues
This release also reduces the dependency on MSOnline but does not yet fully replace it with AzureAD (Issue #19).

---
This change is [Reviewable](https://reviewable.io/reviews/microsoft/office365dsc/104)

### CODE REVIEW

**Comment by NikCharlebois** [comment]
Please update your fork/branch with latest changes. It will help resolve some of the tests that fail.

File: Modules/Office365DSC/Modules/O365DSCReverseGUI.psm1 [hunk 1]
```diff
@@ -391,6 +391,15 @@ function Show-O365GUI
         $txtTenantAdmin.Left = 1060
         $txtTenantAdmin.Width = 175
         $txtTenantAdmin.Font = [System.Drawing.Font]::new($txtTenantAdmin.Font.Name, 10)
+        # Pre-populate the tenant admin box with the global admin username, if we already know it (e.g. from a recent attempt), otherwise try to guess it from the current user's UPN
```
Comment (NikCharlebois) [inline Modules/Office365DSC/Modules/O365DSCReverseGUI.psm1]: I don't think we should try to do this honestly. Chances are that user may be different on each extraction, and for sure the user will never be a windows users. I say we leave that part out from this PR

File: Modules/Office365DSC/Modules/Office365DSCUtil.psm1 [hunk 2]
```diff
@@ -1450,8 +1450,8 @@ function Test-SPOServiceConnection
     )
     $VerbosePreference = 'SilentlyContinue'
     $WarningPreference = "SilentlyContinue"
-    Write-Verbose -Message "Verifying the LCM connection state to SharePoint Online"
-    $catch = Connect-SPOService -Url $SPOCentralAdminUrl -Credential $GlobalAdminAccount
+    Write-Verbose "Verifying the LCM connection state to SharePoint Online"
```
Comment (NikCharlebois) [inline Modules/Office365DSC/Modules/Office365DSCUtil.psm1]: We recently standardized all the Verbose output. Please us -Message

File: Modules/Office365DSC/Modules/Office365DSCUtil.psm1 [hunk 3]
```diff
@@ -1462,16 +1462,16 @@ function Test-PnPOnlineConnection
     (
         [Parameter(Mandatory = $true)]
         [System.String]
-        $SiteUrl,
+        $SPOCentralAdminUrl,
```
Comment (NikCharlebois) [inline Modules/Office365DSC/Modules/Office365DSCUtil.psm1]: This has to be SiteURL, not SPOCentralAdminUrl. PnP let's you establish a connection with any site collection, not just Central Admin

File: Modules/Office365DSC/Modules/Office365DSCUtil.psm1 [hunk 4]
```diff
@@ -1462,16 +1462,16 @@ function Test-PnPOnlineConnection
...
 
         [Parameter(Mandatory = $true)]
         [System.Management.Automation.PSCredential]
         $GlobalAdminAccount
     )
     $VerbosePreference = 'SilentlyContinue'
     $WarningPreference = "SilentlyContinue"
-    Write-Verbose -Message "Verifying the LCM connection state to SharePoint Online with PnP"
-    $catch = Connect-PnPOnline -Url $SiteUrl -Credentials $GlobalAdminAccount
+    Write-Verbose "Verifying the LCM connection state to SharePoint Online with PnP"
```
Comment (NikCharlebois) [inline Modules/Office365DSC/Modules/Office365DSCUtil.psm1]: Standardized output to use -Message

File: Modules/Office365DSC/Modules/Office365DSCUtil.psm1 [hunk 5]
```diff
@@ -1715,13 +1715,33 @@ function Get-UsersLicences
...
-    Write-Verbose -Message "Store all users licences information in Global Variable for futur usage."
+
+    Write-Verbose "Store all users licenses information in Global Variable for future usage."
+
     #Store information to be able to check later if the users is correctly licences for features.
-    if ($null -eq $Global:UsersLicences)
+    if ($null -eq $Global:UsersLicenses)
     {
-        $Global:UsersLicences = Get-MsolUser -All | Select-Object UserPrincipalName, isLicensed, Licenses
+<#         $Global:UsersLicenses = @()
```
Comment (NikCharlebois) [inline Modules/Office365DSC/Modules/Office365DSCUtil.psm1]: Why are we changing the logic here?

File: Modules/Office365DSC/Office365DSC.psd1 [hunk 6]
```diff
@@ -53,8 +53,9 @@
   RequiredModules = @(@{ModuleName = "ReverseDSC"; ModuleVersion = "1.9.4.4"; },
                       @{ModuleName = "MSOnline"; ModuleVersion = "1.1.183.17"; },
                       @{ModuleName = "SharePointPnPPowerShellOnline"; ModuleVersion = "3.5.1901.0"; },
-                      @{ModuleName = "Microsoft.Online.SharePoint.PowerShell"; ModuleVersion = "16.0.8316.0"; },
+                      @{ModuleName = 'Microsoft.Online.SharePoint.PowerShell'; ModuleVersion = "16.0.8316.0"; },
```
Comment (NikCharlebois) [inline Modules/Office365DSC/Office365DSC.psd1]: Double quotes

File: Modules/Office365DSC/Office365DSC.psd1 [hunk 7]
```diff
@@ -53,8 +53,9 @@
...
                       @{ModuleName = "MicrosoftTeams"; ModuleVersion = "1.0.0"; },
+                      @{ModuleName = 'MSCloudLoginAssistant'; ModuleVersion = "0.4.0"; },
```
Comment (NikCharlebois) [inline Modules/Office365DSC/Office365DSC.psd1]: Double Quotes

File: Modules/Office365DSC/Modules/O365DSCReverseGUI.psm1 (hunk 1 above)
Comment (brianlala) [reply Modules/Office365DSC/Modules/O365DSCReverseGUI.psm1]: OK, removing.

File: Modules/Office365DSC/Modules/Office365DSCUtil.psm1 (hunk 2 above)
Comment (brianlala) [reply Modules/Office365DSC/Modules/Office365DSCUtil.psm1]: Done.

File: Modules/Office365DSC/Modules/Office365DSCUtil.psm1 (hunk 3 above)
Comment (brianlala) [reply Modules/Office365DSC/Modules/Office365DSCUtil.psm1]: Done.

File: Modules/Office365DSC/Modules/Office365DSCUtil.psm1 (hunk 4 above)
Comment (brianlala) [reply Modules/Office365DSC/Modules/Office365DSCUtil.psm1]: Agree and done.

File: Modules/Office365DSC/Modules/Office365DSCUtil.psm1 (hunk 5 above)
Comment (brianlala) [reply Modules/Office365DSC/Modules/Office365DSCUtil.psm1]: The idea was to move away from using MSOL, but IIRC I found this challenging to do as the output isn't the same using AzureAD, hence the convoluted logic. WDYT?

File: Modules/Office365DSC/Office365DSC.psd1 (hunk 6 above)
Comment (brianlala) [reply Modules/Office365DSC/Office365DSC.psd1]: Agree and done.

File: Modules/Office365DSC/Modules/Office365DSCUtil.psm1 (hunk 5 above)
Comment (NikCharlebois) [reply Modules/Office365DSC/Modules/Office365DSCUtil.psm1]: I say we leave it as is for the moment. Remove the comments and we should be all go to go. Will do a final round of integration tests once you removed them.

Comment (brianlala) [reply Modules/Office365DSC/Modules/Office365DSCUtil.psm1]: OK done.

**Pull request merged**